src/adapter/src/catalog/state.rs

```diff
@@ -2264,60 +2264,134 @@ impl CatalogState {
         }
     }
 
-    /// Update in-memory catalog state.
+    /// Update in-memory catalog state from a list of updates made to the durable catalog state.
+    ///
+    /// This is meant specifically for bootstrapping because it does not produce builtin table
+    /// updates. The builtin tables need to be loaded before we can produce builtin table updates
+    /// which creates a bootstrapping problem.
+    // TODO(jkosh44) It is very IMPORTANT that per timestamp, the updates are sorted retractions
+    // then additions. Within the retractions the objects should be sorted in reverse dependency
+    // order (objects->schema->database, replica->cluster, etc.). Within the additions the objects
+    // should be sorted in dependency order (database->schema->objects, cluster->replica, etc.).
+    // Objects themselves also need to be sorted by dependency order, this will be tricky but we can
+    // look at the existing bootstrap code for ways of doing this. For now we rely on the caller
+    // providing objects in dependency order.
     #[instrument]
-    pub(crate) fn apply_updates(&mut self, updates: Vec<StateUpdate>) {
-        fn apply<K, V>(map: &mut BTreeMap<K, V>, key: K, value: V, diff: Diff)
+    pub(crate) fn apply_updates_for_bootstrap(&mut self, updates: Vec<StateUpdate>) {
+        for StateUpdate { kind, diff } in updates {
+            assert_eq!(
+                diff, 1,
```

Comment:
The comment above says that retractions must be first. The existence of retractions implies that a diff of `-1` could be present. Am I missing something?

Reply:
No, I think it's just a little confusing in it's current intermediate state. The goal is to reuse `apply_update` in both bootstrapping and catalog transactions. However, I'm starting with bootstrapping because it's much easier to do it piece by piece in bootstrapping. However, it just so happens that catalog snapshots during bootstrap are always consolidated and therefore will always have a positive diff. So unfortunately, right now, the retraction code is all technically dead code. The comment is more appropriate for the usage of `apply_update` within catalog transactions, which will have retractions, but doesn't exist yet. I just wanted to write this all down now while it's in my head so that I don't forget. I'm hoping that once it hit's it's final state, things become more clear.

The assert here isn't really necessary, so I can remove it if that makes things more clear.